File: LinkedList/Singly/insertBefore.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  prepend(data) {
    const newNode = new Node(data);

    newNode.next = this.head;
    this.head = newNode;
    this.length++;

    if (this.length === 1) {
      this.tail = newNode;
    }
  }

  insertBefore(target, data) {
    if (!this.head) {
      throw new Error("List is Empty");
    }

    let newNode = new Node(data);

    let current = this.head;

    if (this.head.data === target) {
      newNode.next = this.head;
      this.head = newNode;
      this.length++;
      return;
    }

    while (current.next !== null && current.next.data !== target) {
      current = current.next;
    }

    if (!current.next) {
      throw new Error("Target not found");
    }

    newNode.next = current.next;
    current.next = newNode;
    this.length++;
  }

  print() {
    let current = this.head;

    let result = "";
    while (current !== null) {
      result += `${current.data}` + `${current.next !== null ? " -> " : ""}`;
      current = current.next;
    }
    return result;
  }
}

const node = new LinkedList(10);

node.prepend(20);
node.prepend(30);
node.prepend(40);

console.log(node.print());

node.insertBefore(30, 25);
console.log(node.print());

node.insertBefore(40, 76);
console.log(node.print());

File: LinkedList/Singly/Python/Insertion/insertBefore.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def prepend(self,data):

        node = Node(data)
        node.next = self.head
        self.head = node
        self.length += 1

        if self.tail is None:
            self.tail = node

        return True
    

    def insertBefore(self, target, data):

        current = self.head

        node = Node(data)

        if(self.head.data == target):
            node.next = self.head
            self.head = node
            self.length += 1

            return True
        
        while current != None and current.next.data != target:
            current = current.next

        if current.next == None:
            return False
        
        node.next = current.next
        current.next = node
        self.length += 1

        return True
    
    def print(self):

        current = self.head
        result = ""
        while current:

            result += f"{current.data}" + ("->" if current.next else "")
            current = current.next

        print(result)
    
if __name__ == "__main__":

    linked_list = LinkedList()

    linked_list.prepend(10)
    linked_list.prepend(20)
    linked_list.prepend(30)
    linked_list.prepend(40)

    linked_list.insertBefore(30,50)
    linked_list.insertBefore(20,70)

    linked_list.print()

            
            

